### ISSUE TITLE
External Labels Overwrite existing Labels in Thanos

### ISSUE THREAD

**Issue by ebini** [thanos-io/thanos#1579]
Hi,

i’m using a prometheus federation installation which scrapes the data from 3 Prometheus instances. (Don't ask why, this is at the moment necessary)
 
External Labels are:

**Prometheus Federate:**
    _app=„federate“
    env=„prod“_

**Prometheus 1**
  _app=„abc“
  env=„dev“_

**Prometheus** 2
  _app=„abc“
  env=„test“_

**Prometheus 3**
  _app=„abc“
  env=„prod“_

In the Prometheus Federation my queries locks fine:
for example on Prometheus (Federate) the query
`mymetric{env=„prod“}
`
returns the correct metric
mymetric{app=„ abc“,env=„prod“} = 0

But on Thanos Query Gateway the Announced LabelSets of the Prometheus Federate is    app=„federate“ and env=„prod“.

In it seems to overwrite the existing Labels in the metrics
So if i do the same query:
`mymetric{env=„prod“}
`
i now get 3 metrics, all with the same labels:
mymetric{app=„federate“,env=„prod“} = 0
mymetric{app=„federate“,env=„prod“} = 0
mymetric{app=„federate“,env=„prod“} = 0

It seems Thanos „overwrites“ the existing Labels with the accounced ones.
I don't know if this is a bug, or i have done something wrong.
Perhaps you can help
(Note, at the moment i have to scrape a prometheus federation installation and can not change to prometheus with thanos sidecar installed)

Thanks in advance

**Comment by krasi-georgiev** [comment]
Maybe try playing with the dedup labels in the querier.

https://github.com/thanos-io/thanos/blob/master/docs/components/query.md#deduplication

**Comment by GiedriusS** [comment]
Hi, perhaps you need to specify `--query.replica-label=app` on Thanos Query?

**Comment by ebini** [comment]
Thanks, but it doesn't change anything.
It still seems that the direkt query to Prometheus isn't the same than using Thanos.

**Comment by MarcMielke** [comment]
same problem, looks like the ingesting prometheus instance wins and overwrites the external label(s) already set in the prometheus instances being scraped from the federation endpoint. `honor-labels` is ignored as it seems. here is what I tried (where `cluster` is the external label):
```
    - job_name: monitoring/gke-preprod
      honor_timestamps: true
      honor_labels: true
      scrape_interval: 60s
      scrape_timeout: 30s
      metrics_path: /federate
      scheme: http
      params:
        match[]:
        - '{__name__=~".+"}'
      static_configs:
      - targets:
        - xx.xx.xx.xx:9090
        labels:
          cluster: gke-preprod
      metric_relabel_configs:
      - separator: ;
        regex: cluster
        target_label: cluster
        replacement: gke-preprod
        action: replace
```

**Comment by QianChenglong** [comment]
This problem is a bit annoying, hopefully it can be fixed.

**Comment by ashleyprimo** [comment]
Hi :wave: - has anyone found a fix for this?

**Comment by ldb** [comment]
Hello,
we are having the same problem:

A federated Prometheus instance scrapes local Kubernetes pods and sets the external label `env=dev-k8s`.
However, it also scrapes metrics from EC2 instances which all contain the label `env=dev-ec2`.

When querying that instance directly, we can find the metrics using the label `env=dev-ec2`. However, when using the Thanos Query Frontend or Querier, all metrics have the label `env=dev-k8s`.

According to [Prometheus Documentation](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config) this should not happen:

> Note that any globally configured "external_labels" are unaffected by this
> setting. In communication with external systems, they are always applied only
> when a time series does not have a given label yet and are ignored otherwise.

In other words, external labels should only be added to a metric if they do not already exist, or not?

**Comment by ebini** [comment]
in my opinion still an issue

**Comment by pahaeanx** [comment]
This is still an issue in 0.31.0 and the 0.32rc -- thanos sidecar seems to always "apply" `external_labels`, even when a timeseries already has such a label set. This makes querying locally and via thanos give different results as explained here https://github.com/thanos-io/thanos/issues/1579#issuecomment-1102846336

**Comment by GiedriusS** [comment]
Yes, this is by design and I don't think there's anything to do here. Otherwise, the deduplication cannot work correctly. Please see: https://github.com/thanos-io/thanos/pull/6564, https://github.com/thanos-io/thanos/issues/6257#issuecomment-1653169915.

**Comment by pahaeanx** [comment]
Thank your for the quick reply! I'm still a bit confused however, this seems a valid use case.

This is a rather annoying problem in some situations and I think there's no easy workaround, unfortunately.

**Comment by fpetkovski** [comment]
Can you change your external labels to not have conflicts with series labels?

**Comment by pahaeanx** [comment]
Unfortunately, no. These external labels are used as selectors. 

In our case it happens with a setup like this: 

```yaml
external_labels:
  cluster: A
  replica: 1
  
scrape_configs:
- job_name: 'a-single-job-for-some-other-cluster'
  static_configs:
  - targets: ['target:9100']
    labels:
      cluster: B
      
... a bunch of scrape jobs for targets in our cluster (A).
```
As this Prometheus only has (literally) one job scraping another "cluster" (using the term lightly here), this seems like the perfect solution. And indeed, querying metrics on the Prometheus itself yields the correct `cluster=B` label. Using thanos yields `cluster=A` as the sidecar indiscriminately applies the external label(s).

**Comment by GiedriusS** [comment]
Unfortunately, as explained ealier, it works as intended. External labels are used for routing i.e. reduce fan-out scope and they always take precedence. The fix is to not reuse the same labels as external & series labels.
